types/tx.go

```diff
@@ -102,6 +94,232 @@ func ToTxs(txs [][]byte) Txs {
+	// and once by sorting the set of the added, removed, and unmodified transactions indexes,
+	// which, when combined, comprise the complete list of transactions.
+	//
+	// The original list is iterated once and the modified list is iterated multiple times,
+	// one time alongside each of the 3 indexes for a total of 4 iterations of the modified list.
+	// Asymptotically, this yields a total runtime of O(N*log(N) + N + 2*M*log(M) + 4*M),
+	// in the input size of the original list and the input size of the new list respectively.
+	// A 2 * M performance gain is possible but iterating all of the indexes simultaneously
+	// alongside the original list, but the multiple iterations were preferred to be more
+	// readable and maintainable.
+
+	// Sort a copy of the complete transaction slice so we can check for
+	// duplication. The copy is so we do not change the original ordering.
+	// Only the slices are copied, the transaction contents are shared.
+	allCopy := make([]Tx, len(t.all))
+	copy(allCopy, t.all)
+	sort.Sort(Txs(allCopy))
+
+	var size int64
+	for i, cur := range allCopy {
+		size += int64(len(cur))
+		if size > maxSizeBytes {
+			return fmt.Errorf("transaction data size %d exceeds maximum %d", size, maxSizeBytes)
+		}
+
+		// allCopy is sorted, so any duplicated data will be adjacent.
+		if i+1 < len(allCopy) && bytes.Equal(cur, allCopy[i+1]) {
+			return fmt.Errorf("found duplicate transaction with hash: %x", cur.Hash())
+		}
+	}
+
+	// create copies of each of the action-specific indexes so that order of the original
+	// indexes can be preserved.
+	addedCopy := make([]Tx, len(t.added))
+	copy(addedCopy, t.added)
+	sort.Sort(Txs(addedCopy))
+
+	removedCopy := make([]Tx, len(t.removed))
+	copy(removedCopy, t.removed)
+	sort.Sort(Txs(removedCopy))
+
+	unmodifiedCopy := make([]Tx, len(t.unmodified))
+	copy(unmodifiedCopy, t.unmodified)
+	sort.Sort(Txs(unmodifiedCopy))
+
+	// make a defensive copy of otxs so that the order of
+	// the caller's data is not altered.
+	otxsCopy := make([]Tx, len(otxs))
+	copy(otxsCopy, otxs)
+	sort.Sort(Txs(otxsCopy))
+
+	if ix, ok := containsAllTxs(otxsCopy, unmodifiedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", unmodifiedCopy[ix].Hash())
+	}
+
+	if ix, ok := containsAllTxs(otxsCopy, removedCopy); !ok {
+		return fmt.Errorf("new transaction incorrectly marked as removed, transaction hash: %x", removedCopy[ix].Hash())
+	}
+	if ix, ok := containsNoneTxs(otxsCopy, addedCopy); !ok {
+		return fmt.Errorf("existing transaction incorrectly marked as added, transaction hash: %x", addedCopy[ix].Hash())
+	}
+	return nil
+}
+
+// containsNoneTxs checks that list a contains none of the transactions in list
+// b. If a match is found, the index in b of the matching transaction is returned.
+// Both lists must be sorted.
+func containsNoneTxs(a, b []Tx) (int, bool) {
+	aix, bix := 0, 0
+	for ; aix < len(a); aix++ {
+	LOOP:
```

Comment:
Do we not have to reset `bix = 0` for the next element of `a`?

Reply:
hmm, so, we're leveraging the fact that both lists are sorted. Each of the lists is increasing, which means that we don't have to assume that we may have missed a value of `b` that is present in `a` and reset bix to 0. We advance either iterator until the value under it is greater than or equal to the value under the other iterator. If it's equal, we terminate since we've found a match. If it's greater, we advance the other iterator. If we were to miss a value of `b` that was present in `a` then we would have had to have increased `b`'s iterator past the value in `b` despite their being a matching value in `a`. However, we don't do this because we only advance `b`'s iterator when its value is less than that of `a`'s iterator. Let me know if you don't buy this and I can think on it a bit more.